app/_src/gateway/plugin-development/wasm/filter-configuration.md

```diff
@@ -0,0 +1,226 @@
+---
+title: Proxy-Wasm Filter Configuration
+book: plugin_dev
+chapter: 12
+---
+
+## Proxy-Wasm Filter Configuration
+
+### Purpose
+
+The [Proxy-Wasm specification](https://github.com/proxy-wasm/spec) defines
+filter runtime configuration only as an array of bytes. The finer details of
+configuration-handling (encoding, serialization, validation) are left to the
+filter to implement.
+
+As a result, Proxy-Wasm implementations have no means of validating
+configuration until runtime, when a filter instance is created. In
+{{site.base_gateway}}, executing a filter with an invalid configuration will
+trigger a 500 error, as Kong has no choice but to assume that it is not safe to
+continue processing the request.
+
+This presents a challenge for filter deployment: how do operators prevent users
+from creating improperly-configured filters?
+
+### JSON Schema for Filter Configuration
+
+Filters that take configuration in JSON form can optionally make use of 
+{{site.base_gateway}}'s [JSON schema](https://json-schema.org/) validation.
+
+This is accomplished by providing a JSON schema in a filter metadata file
+adjacent to the filter's bytecode within the `wasm_filters_path` directory:
+
+```
+/path/to/wasm-filters
+├── my-filter.wasm
+├── my-filter.meta.json
+├── other-filter.wasm
+└── other-filter.meta.json
+```
+
+Configuration schema is supplied in the `config_schema` metadata field. Example:
+
+```json
+{
+  "config_schema": {
+    "type": "object",
+    "properties": {
+      "my-property": {
+        "type": "string"
+      }
+    },
+    "required": [
+      "my-property"
+    ]
+  }
+}
+```
+
+The presence of a schema allows Kong to interpret the filter configuration as
+JSON instead of treating it as an opaque string when reading and writing request
+and response bodies.
+
+Without a schema, JSON filter configuration must be serialized as a string in
+API requests:
+
+```console
+$ cat <<'EOF' | http POST :8001/services/test/filter-chains
```

Comment:
```suggestion
cat <<'EOF' | http POST :8001/services/test/filter-chains
```
Can you please convert the examples to cURL? We've stopped using HTTPie in the docs because it's an extra tool for users to install (we've had complaints of customers getting confused and thinking that it's a required tool, which is a problem when it comes to large companies with severe restrictions on software).

Reply:
Alternative idea... so, all of the HTTP requests in the article are informational/demonstrative ("look at what happens when XYZ") and not meant to be step-by-step instructions that a user will copy+paste.

In light of that, what do you think about formatting the request/response as plain text instead of providing a shell snippet? Something like...

```
POST /services/test/filter-chains HTTP/1.1
Host: kong
User-Agent: curl/8.0.1
Content-Type: application/json
Accept: application/json
Content-Length: 211

{
  "name": "my-filter-chain",
  "filters": [
    {
      "name": "set-response-header",
      "config": {
        "header-name": "X-My-Custom-Header",
        "header-value": "Hello, Wasm!"
      }
    }
  ]
}
```

```
HTTP/1.1 201 Created
Access-Control-Allow-Credentials: true
Access-Control-Allow-Origin: *
Connection: keep-alive
Content-Length: 348
Content-Type: application/json; charset=utf-8
Date: Mon, 06 Nov 2023 00:21:30 GMT
Server: kong/3.5.0
X-Kong-Admin-Latency: 23

{
    "created_at": 1699230090,
    "enabled": true,
    "filters": [
        {
            "config": {
                "header-name": "X-My-Custom-Header",
                "header-value": "Hello, Wasm!"
            },
            "enabled": true,
            "name": "set-response-header"
        }
    ],
    "id": "da8659e3-db85-435f-ba62-f95ec7615ddc",
    "name": "my-filter-chain",
    "route": null,
    "service": {
        "id": "992c4491-9962-48ff-8d5d-249ba82848d1"
    },
    "tags": null,
    "updated_at": 1699230090
}
```

Reply:
I like that idea, let's go with it.